support/node-health-check/src/main.rs

```diff
@@ -0,0 +1,107 @@
+use std::convert::Infallible;
+use std::sync::atomic::AtomicBool;
+use std::sync::atomic::Ordering;
+use std::{sync::Arc, time::Duration};
+
+use anyhow::Result;
+use reqwest::StatusCode;
+use serde::{Deserialize, Serialize};
+use structopt::StructOpt;
+use warp::Filter;
+
+/// Health struct returned by the RPC
+#[derive(Debug, PartialEq, Serialize, Deserialize, Default)]
+#[serde(rename_all = "camelCase")]
+pub struct Health {
+    /// Number of connected peers
+    pub peers: usize,
+    /// Is the node syncing
+    pub is_syncing: bool,
+    /// Should this node have any peers
+    /// Might be false for local chains or when running without discovery.
+    pub should_have_peers: bool,
+}
+
+#[derive(Debug, StructOpt, Clone)]
+#[structopt(
+    name = "node-health-check",
+    about = "Sidecar for substrate node to report health compatible with AWS ALB."
+)]
+struct Opt {
+    /// Debug
+    #[structopt(short, long)]
+    debug: bool,
+
+    /// Set poll_interval
+    #[structopt(short = "i", default_value = "5")]
+    poll_interval: u64,
+
+    /// Set minimum peers
+    #[structopt(default_value = "2")]
+    min_peers: usize,
+
+    /// Node rpc endpoint
+    #[structopt(short, long, env, default_value = "http://localhost:9933/health")]
+    node_rpc_endpoint: String,
+
+    // The port to listen on
+    #[structopt(short, long, default_value = "9955")]
+    port: u16,
+}
+
+async fn poller(opts: Opt, healthy: Arc<AtomicBool>) {
+    loop {
+        let resp = async {
+            reqwest::get(&opts.node_rpc_endpoint)
+                .await?
```

Comment:
I'm unclear as to the behavior of `?` here. I'd expect it to return from the function, but it seems to be returning from the block?

If it returns from the function, then your poller would stop if the request fails and would only hit the `Err` case later if specifically loading the JSON failed. So I think that if the node is unreachable, the poller would stop and everything would look fine on AWS (despite it clearly not being fine).

Reply:
Apparently, according to this: https://rust-lang.github.io/async-book/07_workarounds/02_err_in_async_blocks.html. the `?` operator returns from the async block.

Probably good to explicitly mark the return type of this function as a never type, `!` when the `never_type` feature is stabilized. Or an enum with no variants until then. `enum Never {}`. That makes it clear that it's impossible for the function to return.

Reply:
Yes this is a trick I started using, and see being used in the wild as an alternative to a try block, which is also an unstable feature in rust. It's just to keep the code a bit cleaner/readable, so I can use the double ? chain, and not have to do nested matches or nested ifs.

Reply:
Can you make the return type of the `poller` fn `Never` to make this more explicit?

Reply:
Yes, I like it! I already stripped the Result from the `poller` to communicate no `?` could be used outside the block I was using it. This makes it even more explicit for future implementors.